### PR TITLE
Feature/phoenix upgrade

### PR DESCRIPTION
**Pull request by roman-khomenko-dev**
Upgrade phoenix version to 1.7.1
Rewrite messages functionality using liveview streams
Setting up releases deploy to fly.io
Change tests according to messages index changes

### CODE REVIEW

File: lib/chatter_web/live/chat_live/index.ex [hunk 1]
```diff
@@ -6,13 +6,29 @@ defmodule ChatterWeb.ChatLive.Index do
...
+  end
+
+  @impl true
+  def handle_info({Message, {:message, :updated}, message}, socket) do
+    {:noreply, stream_insert(socket, :messages, message, at: -1)}
+  end
+
+  @impl true
+  def handle_info({Message, {:message, :removed}, message}, socket) do
+    {:noreply, stream_delete(socket, :messages, message)}
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:31]: do we remove ?

File: lib/chatter_web/live/chat_live/index.ex [hunk 2]
```diff
@@ -42,24 +58,31 @@ defmodule ChatterWeb.ChatLive.Index do
 
   @impl true
   def handle_event("search", %{"search" => params}, socket) do
-    with {:ok, search} <- Search.create(params) do
+    with {:ok, search} <- Search.create(params),
+      search_messages <- user_shown_messages({search, socket}) do
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:62]: do you need full socket here ?

File: lib/chatter_web/live/chat_live/index.ex [hunk 3]
```diff
@@ -42,24 +58,31 @@ defmodule ChatterWeb.ChatLive.Index do
...
+      broadcast_filtered_message(search_messages, socket.assigns.username)
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:63]: `socket.assigns.username` can be pattern matched in args

File: lib/chatter_web/live/chat_live/index.ex [hunk 4]
```diff
@@ -42,24 +58,31 @@ defmodule ChatterWeb.ChatLive.Index do
...
+    socket =
+      socket
+      |> assign_filter_option(option)
+      |> activate_show("show_menu")
+
+    {search, socket}
+    |> user_shown_messages()
+    |> broadcast_filtered_message(socket.assigns.username)
+
+    {:noreply, socket}
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:85]: might be a good cadidate for `{:noreply_continue, ....}`

File: lib/chatter_web/live/chat_live/messages_component.ex [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  use ChatterWeb, :live_component
+
+  def render(assigns) do
+    ~H"""
+    <div id="messages" phx-update="stream">
+        <div :for={message <- @messages} class="card col-10 mx-auto d-block my-5 shadow mb-5 bg-white rounded" id={elem(message,0)}>
+            <div class="card-body">
+                <p class="card-text">
+                    <%= elem(message, 1).text %>
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/messages_component.ex:14]: I doubt FE dev can understand this

File: lib/chatter_web/live/chat_live/index.ex (hunk 2 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex]: no, changed `user_shown_messages` params  to {search, filter_option}

File: lib/chatter_web/live/chat_live/index.ex (hunk 3 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex:63]: changed to gather username with pattern matching

File: lib/chatter_web/live/chat_live/index.ex [hunk 6]
```diff
@@ -42,24 +58,31 @@ defmodule ChatterWeb.ChatLive.Index do
...
-       |> assign(messages: get_filtered_messages(search))
        |> assign(search: Search.change_search(%Search{}, params))
        |> activate_show("show_menu")}
     end
   end
 
   @impl true
   def handle_event("filter_option", %{"option" => option} = _params, socket) do
-    option = String.to_atom(option)
+    {option, search} = {String.to_atom(option), apply_search(socket)}
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex]: i see, changed to use safe String.to_existing_atom instead

File: lib/chatter_web/live/chat_live/index.html.heex [hunk 7]
```diff
@@ -18,29 +18,8 @@
...
-                            <%= Bootstrap.Icons.star_fill()%>
-                        </a>
-                        <%= length(message.likes) %>
-                    </span>
-                </small>
-            </div>
-        </div>
-    <% end %>
+
+    <.live_component module={MessagesComponent} id="messages-container" username={@username} messages={@streams.messages} />
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.html.heex]: sure

File: lib/chatter_web/live/chat_live/messages_component.ex (hunk 5 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/messages_component.ex]: changed to more clear way

File: lib/chatter_web/live/chat_live/index.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex:31]: in this case handle_info used to responde for broadcast sygnal that message need to be removed from socket streams for this user

File: lib/chatter/message_filter.ex [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+  alias Chatter.{MessageAgent, Search}
+
+  def filter_by_params({search, nil = _filter_option}), do: get_search_messages(search)
+
+  def filter_by_params({search, filter_option}) do
+    search_messages = get_search_messages(search)
+    filter_option
+      |> filter_by_option()
+      |> Enum.filter(fn message -> message in search_messages end)
```
Comment (sergey-lukianov) [inline lib/chatter/message_filter.ex:14]: add  in your git actions pipeline
```
	mix compile --warnings-as-errors
	mix format --check-formatted
	mix credo --strict
```

File: lib/chatter/message_filter.ex [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  defp filter_likes(message, %{likes_option: option, likes: likes_count} = _search) do
+    case option do
+      ">=" -> Enum.count(message.likes) >= likes_count
+      "<=" -> Enum.count(message.likes) <= likes_count
+      "=" -> Enum.count(message.likes) == likes_count
+    end
+  end
+
+  defp filter_with_likes_who_like do
+    all_likes = MessageAgent.get_all_likes()
```
Comment (sergey-lukianov) [inline lib/chatter/message_filter.ex:46]: hardcoding agent usage is a bad thing
it is better to pass it as arguments

File: lib/chatter/message_filter.ex [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    Enum.filter(MessageAgent.get(), fn message ->
+      if Enum.count(message.likes) > 0 && message.author in all_likes, do: message
```
Comment (sergey-lukianov) [inline lib/chatter/message_filter.ex:48]: `count` -> `empty?`

File: lib/chatter_web/live/chat_live/index.ex [hunk 11]
```diff
@@ -68,46 +91,48 @@ defmodule ChatterWeb.ChatLive.Index do
...
+      |> Enum.member?(message)
 
-  def assign_filter_option_messages(socket) do
-    assign_search_messages(socket)
+    {is_member, message, socket} |> created_member_response()
   end
 
+  defp created_member_response({false, _message, socket}), do: {:noreply, socket}
+
+  defp created_member_response({true, message, socket}), do: {:noreply, stream_insert(socket, :messages, message, at: 0)}
```
Comment (sergey-lukianov) [inline lib/chatter_web/live/chat_live/index.ex:126]: returning `:noreply` within a function you restrict use of composition

File: lib/chatter/message_filter.ex (hunk 9 above)
Comment (roman-khomenko-dev) [reply lib/chatter/message_filter.ex]: right, messages data moved to params

File: lib/chatter/message_filter.ex (hunk 10 above)
Comment (roman-khomenko-dev) [reply lib/chatter/message_filter.ex]: changed

File: lib/chatter_web/live/chat_live/index.ex (hunk 11 above)
Comment (roman-khomenko-dev) [reply lib/chatter_web/live/chat_live/index.ex]: `:noreply` moved to `handle_info` returns

**Pull request merged**